src/Microsoft.AspNet.Diagnostics/ErrorPageMiddleware.cs

```diff
@@ -182,31 +188,104 @@ private StackFrame StackFrame(Chunk line)
+                    // "--- End of stack trace from previous location where exception from thrown ---"
+                    string.IsNullOrEmpty(function) ? line.ToString() : function,
+                    file: string.Empty,
+                    startLineNumber: 0,
+                    endLineNumber: 0);
+            }
+            else
+            {
+                return GetStackFrame(function, file, lineNumber, lineNumber);
+            }
         }
 
-        private StackFrame LoadFrame(string function, string file, int lineNumber)
+        // make it internal to enable unit testing
+        internal StackFrame GetStackFrame(string function, string file, int startLineNumber, int endLineNumber)
         {
-            var frame = new StackFrame { Function = function, File = file, Line = lineNumber };
+            var frame = new StackFrame { Function = function, File = file, Line = startLineNumber };
+
+            if (string.IsNullOrEmpty(file))
+            {
+                return frame;
+            }
+
+            IEnumerable<string> lines = null;
             if (File.Exists(file))
             {
-                var code = File.ReadLines(file);
-                ReadFrameContent(frame, code, lineNumber, lineNumber);
+                lines = File.ReadLines(file);
             }
+            else
+            {
+                // Handle relative paths and embedded files
+                var fileInfo = _options.FileProvider.GetFileInfo(file);
+                if (fileInfo.Exists)
+                {
+                    // ReadLines doesn't accept a stream. Use ReadLines as its more efficient
+                    // relative to reading lines via stream reader
+                    if (!string.IsNullOrEmpty(fileInfo.PhysicalPath))
+                    {
+                        lines = File.ReadLines(fileInfo.PhysicalPath);
+                    }
+                    else
+                    {
+                        lines = ReadLines(fileInfo);
+                    }
+                }
+            }
+
+            if (lines != null)
+            {
+                ReadFrameContent(frame, lines, startLineNumber, endLineNumber);
+            }
+
             return frame;
         }
 
-        private void ReadFrameContent(StackFrame frame,
-                                      IEnumerable<string> code,
-                                      int startLineNumber,
-                                      int endLineNumber)
+        private void ReadFrameContent(
+            StackFrame frame,
+            IEnumerable<string> allLines,
+            int errorStartLineNumberInFile,
+            int errorEndLineNumberInFile)
+        {
+            // Get the line boundaries in the file to be read and read all these lines at once into an array.
+            // This will cause the backing stream to be closed.
+            var preErrorLineNumberInFile = Math.Max(errorStartLineNumberInFile - _options.SourceCodeLineCount, 1);
```

Comment:
`_options.SourceCodeLineCount` is that the total source code lines? If so, doesn't this change the behavior to always have the entire file as the code block?